clarity/src/vm/ast/parser_v2/lexer/mod.rs

```diff
@@ -0,0 +1,1957 @@
+pub mod error;
+pub mod token;
+
+use crate::vm::{
+    diagnostic::{DiagnosableError, Level},
+    representations::Span,
+};
+use std::{char, str::Chars};
+use token::{PlacedToken, Token};
+
+use crate::vm::ast::parser_v2::lexer::error::LexerError;
+
+use self::error::PlacedError;
+
+pub struct Lexer<'a> {
+    input: Chars<'a>,
+    next: char,
+    offset: usize,
+    pub line: usize,
+    pub column: usize,
+    pub last_line: usize,
+    pub last_column: usize,
+    // While lexing, collect diagnostics and continue on (when possible)
+    pub diagnostics: Vec<PlacedError>,
+    pub success: bool,
+    // Used to report an error for line-endings only on the first instance
+    line_endings: bool,
+    // `fail_fast` mode indicates that the lexer should not report warnings
+    // and should exit on the first error. This is useful for lexing in the
+    // context of a stacks-node, while normal mode is useful for developers.
+    fail_fast: bool,
+}
+
+fn is_separator(ch: char) -> bool {
+    match ch {
+        '(' | ')' | '{' | '}' | ',' | ':' | '\0' | '.' => true,
```

Comment:
What does `\0` separate in v2?  I don't think it's even legal in v1?

Reply:
It should not be allowed in the contract. It's just used as a placeholder for the end of the contract string (see line 82 below):

```rust
        match self.input.next() {
             Some(ch) => {
                 if self.next == '\n' {
                     self.line = self.line + 1;
                     self.column = 0;
                 }
                 self.next = ch
             }
             None => self.next = '\0',
         }
```

Reply:
That's the thing -- I don't think strings in rust are guaranteed to be null-terminated.  There isn't a null byte that terminates the contract payload, for example.

What are we doing to be sure that there is absolutely no way that `\0`, or any other ASCII control code besides `\t` and `\n`, can show up in the code body?  Is this handled by a regex somewhere before the contract is fed into the lexer?

Reply:
Ah, thanks, I see the problem. This technique is not relying on a null-terminator in the rust string, but it is missing a check for `\0` showing up in the code. If it does show up, it would be treated like the end of the file and cause a problem. I'll fix that.

Reply:
Fixed in dbb2664fb